File: exercise/kotlin/day16/src/main/kotlin/system/TaskAssignment.kt
package system

data class Elf(val id: Int, var skillLevel: Int)

class TaskAssignment(private val elves: List<Elf>) {
    private var tasksCompleted: Int = 0

    fun reportTaskCompletion(elfId: Int): Boolean {
        val elf = elves.find { it.id == elfId }
        if (elf != null) {
            tasksCompleted++
            return true
        }
        return false
    }

    fun totalTasksCompleted(): Int = tasksCompleted

    fun elfWithHighestSkill(): Elf? = elves.maxByOrNull { it.skillLevel }

    fun assignTask(taskSkillRequired: Int): Elf? =
        elves.find { it.skillLevel >= taskSkillRequired + 1 }

    fun increaseSkillLevel(elfId: Int, increment: Int) {
        elves.find { it.id == elfId }
            ?.let {
                it.skillLevel += increment
            }
    }

    fun decreaseSkillLevel(elfId: Int, decrement: Int) {
        elves.find { it.id == elfId }
            ?.let {
                if (it.skillLevel - decrement > 0) {
                    it.skillLevel -= decrement
                }
            }
    }

    fun reassignTask(fromElfId: Int, toElfId: Int): Boolean {
        val fromElf = elves.find { it.id == fromElfId }
        val toElf = elves.find { it.id == toElfId }

        return if (fromElf != null && toElf != null && fromElf.skillLevel > toElf.skillLevel) {
            toElf.skillLevel = fromElf.skillLevel
            true
        } else false
    }

    fun listElvesBySkillDescending(): List<Elf> = elves.sortedByDescending { it.skillLevel }
    fun resetAllSkillsToBaseline(baseline: Int) = elves.forEach { it.skillLevel = baseline }
}

File: exercise/scala/day16/src/main/scala/system/TaskAssignment.scala
package system

case class Elf(id: Int, var skillLevel: Int)

class TaskAssignment(private val elves: List[Elf]) {
  private var tasksCompleted: Int = 0

  def reportTaskCompletion(elfId: Int): Boolean = {
    elves.find(_.id == elfId) match {
      case Some(_) =>
        tasksCompleted += 1
        true
      case None => false
    }
  }

  def totalTasksCompleted(): Int = tasksCompleted

  def elfWithHighestSkill(): Option[Elf] = elves.maxByOption(_.skillLevel)

  def assignTask(taskSkillRequired: Int): Option[Elf] =
    elves.find(_.skillLevel >= taskSkillRequired + 1)

  def increaseSkillLevel(elfId: Int, increment: Int): Unit = {
    elves.find(_.id == elfId).foreach { elf =>
      elf.skillLevel += increment
    }
  }

  def decreaseSkillLevel(elfId: Int, decrement: Int): Unit = {
    elves.find(_.id == elfId).foreach { elf =>
      if (elf.skillLevel - decrement > 0) {
        elf.skillLevel -= decrement
      }
    }
  }

  def reassignTask(fromElfId: Int, toElfId: Int): Boolean = {
    (elves.find(_.id == fromElfId), elves.find(_.id == toElfId)) match {
      case (Some(fromElf), Some(toElf)) if fromElf.skillLevel > toElf.skillLevel =>
        toElf.skillLevel = fromElf.skillLevel
        true
      case _ => false
    }
  }

  def listElvesBySkillDescending(): List[Elf] = elves.sortBy(-_.skillLevel)

  def resetAllSkillsToBaseline(baseline: Int): Unit = elves.foreach(_.skillLevel = baseline)
}